src/docker-extract.c

```diff
@@ -417,13 +425,39 @@ int main(int argc, char **argv) {
     }
 
     singularity_message(DEBUG, "Applying whiteouts for tar file %s\n", tarfile);
-    retval = apply_whiteouts(tarfile, rootfs_realpath);
+    retval = apply_whiteouts_and_get_tar_size(tarfile, rootfs_realpath, &exploded_tar_size);
 
     if (retval != 0) {
         singularity_message(ERROR, "Error applying layer whiteouts\n");
         ABORT(255);
     }
 
+    // following logic is to make sure we have enough space on filesystem to explode layers of image
+    if (!exploded_tar_size)
+	singularity_message(WARNING, "Unable to estimate expanded size of archive %s or size is zero.\n", tarfile);
+    
+    singularity_message(DEBUG, "Estimated to-be-exploded archive size to %lu bytes\n", exploded_tar_size);
+
+    char *layer_target_fs = singularity_registry_get("ROOTFS");
+    if (layer_target_fs == NULL)
+	layer_target_fs = "/";
+    
+    if (statvfs(layer_target_fs, &stat) != 0)
+	 fs_avail_space = -1;
+    else
+	fs_avail_space = stat.f_bsize * stat.f_bavail;
+
+    if (fs_avail_space == -1)
+	singularity_message(WARNING, "Unable to obtain free space on %s\n", layer_target_fs);
+
+    singularity_message(DEBUG, "Available /tmp space %lu bytes\n", fs_avail_space);
```

Comment:
Where `/tmp` is hard coded (and below), you should use `layer_target_fs` for the target... like:

`singularity_message(DEBUG, "Available %s space %lu bytes\n", layer_target_fs, fs_avail_space);`